python/cudf/cudf/core/copy_types.py

```diff
@@ -13,128 +14,161 @@
-
-    Parameters
-    ----------
-    column
-        The column to verify
-    nrows
-        The number of rows to verify against
-    nullify
-        Will this gather map be used nullifying out of bounds accesses
-    check_bounds
-        Actually check whether the map is in bounds. Set to False if
-        you know by construction that the map is in bounds.
-
-    Returns
-    -------
-    GatherMap
-        New object wrapping the column bearing witness to its
-        suitability as a gather map for columns with nrows.
-
-    Raises
-    ------
-    IndexError
-        If the column is of unsuitable dtype, or the map is not in bounds.
-    """
-    column = cudf.core.column.as_column(column)
-    if len(column) == 0:
-        # Any empty column is valid as a gather map
-        # This is necessary because as_column([]) defaults to float64
-        # TODO: we should fix this further up.
-        # Alternately we can have an Optional[Column] and handle None
-        # specially in _gather.
-        return GatherMap(
-            cast("NumericalColumn", column.astype(size_type_dtype)),
-            nrows,
-            nullify,
-        )
-    if column.dtype.kind not in {"i", "u"}:
-        raise IndexError("Gather map must have integer dtype")
-    if not nullify and check_bounds:
-        lo, hi = libcudf.reduce.minmax(column)
-        if lo.value < -nrows or hi.value >= nrows:
-            raise IndexError(f"Gather map is out of bounds for [0, {nrows})")
-    return GatherMap(cast("NumericalColumn", column), nrows, nullify)
-
-
-def as_boolean_mask(column: Any, nrows: int) -> BooleanMask:
-    """Turn a column into a boolean mask
-
-    This augments the column with information that it is valid as a
-    boolean mask for columns with a given number of rows
-
-    Parameters
-    ----------
-    column
-        The column to verify
-    nrows
-        the number of rows to verify against
-
-    Returns
-    -------
-    BooleanMask
-        New object wrapping the column bearing witness to its
-        suitability as a boolean mask for columns with matching row
-        count.
-
-    Raises
-    ------
-    IndexError
-        If the column is of unsuitable dtype.
-    """
-    column = cudf.core.column.as_column(column)
-    if column.dtype.kind != "b":
-        raise IndexError("Boolean mask must have bool dtype")
-    if (n := len(column)) != nrows:
-        raise IndexError(
-            f"Column with {n} rows not suitable "
-            f"as a boolean mask for {nrows} rows"
-        )
-    return BooleanMask(cast("NumericalColumn", column))
+    def __init__(self, column: Any, nrows: int):
+        """A representation of a column as a boolean mask.
+
+        This augments the column with information that it is valid as a
+        boolean mask for columns with a given number of rows
+
+        Parameters
+        ----------
+        column
+            The data to turn into a column to then verify
+        nrows
```

Comment:
Why does this class need the `nrows` parameter? We know the number of rows must be `len(column)` or else it's invalid. It doesn't seem useful to require that input, we should just use `len(column)`. This is especially true for the unchecked form.

Reply:
I guess I don't need to store it. The question of requiring it as input is to do with whether you find out early or late that your mask is the wrong length. If you don't require the length of the target (to-be-masked) frame as input, then you get a libcudf `RuntimeError` when `apply_boolean_mask` is called.

Reply:
Note as well, because we don't have dependent types, that the consumers of `BooleanMask` and `GatherMap` do some extra (cheap) runtime validation:

e.g. in `IndexedFrame._gather`
```
    def _gather(
        self,
        gather_map: GatherMap,
        keep_index=True,
    ):
        """Gather rows of frame specified by indices in `gather_map`.

        Maintain the index if keep_index is True.

        This function does no expensive bounds checking, but does
        check that the number of rows of self matches the validated
        number of rows.
        """
        if not gather_map.nullify and len(self) != gather_map.nrows:
            raise IndexError("Gather map is out of bounds")
```

Reply:
removed storing `nrows`.